rfcs/rfc0017.md

```diff
@@ -0,0 +1,149 @@
+flags based on their use, but since numbers will never gain the POK flag, this
+can still be used to detect numbers.
+
+Exposing this information via builtin functions will allow pure perl code to
+accurately serialize values.
+
+## Rationale
+
+While the flags on scalars can be checked via the B module, using it requires
+an understanding of perl internals, and it is slow and uses a fairly large
+amount of memory. Being able to interpret the flags to know string vs number
+is also not entirely obvious.
+
+The functions are intentionally using longer less convenient names, with the
+hope that this will discourage their use for type checks within perl. Inside
+perl, 1 and "1" are meant to be treated equivalently. It is only when
+transferring data outside perl that these are meant to be used.
+
+`created_as` is meant to only tell how a value was created, without implying
+anything further about what the value "is". `created` in this case means any
+modification of the value, as any modification of a value requires creating
+the new, modified value.
+
+## Specification
+
+    created_as_string VALUE
+
+`created_as_string` takes a single scalar argument.
+
+It returns a boolean value representing if the value was originally created as
+a string. It will return false for references, numbers, booleans, and undef.
+Internally, this would be `SvPOK(sv) && !SvIsBOOL(sv)`.
+
+    created_as_number VALUE
+
+`created_as_number` takes a single scalar argument.
+
+It returns a boolean value representing if the value was originally created as
+a number. It will return false for references, strings, booleans, and undef.
+Internally, this would be `SvNIOK(sv) && !SvPOK(sv) && !SvIsBOOL(sv)`.
+
+## Backwards Compatibility
+
+There should be no significant backwards compatibility concerns.
+`created_as_string` and `created_as_number` will be imported only when
+requested. Static analysis tools may need to be updated.
+
+As these functions rely on the changes to how SV flags get set, it isn't
+possible to accurately implement these on older perls. Existing serializers
+use heuristics that will usually give accurate answers. One of these heuristics
+could be used to provide a compatible polyfill.
+
+## Security Implications
+
+Nothing specific predicted.
+
+## Examples
+
+  use builtin qw(created_as_number created_as_string);
+
+  my $value1 = "1";
+  my $string1 = created_as_string $value1; # true
+  my $number1 = created_as_number $value1; # false
+
+  my $value2 = 1;
+  my $string2 = created_as_string $value2; # false
+  my $number2 = created_as_number $value2; # true
+
+  my $value3 = "1";
+  my $used_as_number3 = 0+$value3;
+  my $string3 = created_as_string $value3; # true
+  my $number3 = created_as_number $value3; # false
+
+  my $value4 = 1;
+  my $used_as_string4 = "$value4";
+  my $string4 = created_as_string $value4; # false
+  my $number4 = created_as_number $value4; # true
+
+## Prototype Implementation
+
+None.
```

Comment:
I've started a (currently-draft) core PR for this at https://github.com/Perl/perl5/pull/19524